keps/sig-api-machinery/4568-resilient-watchcache-initialization/README.md

```diff
@@ -0,0 +1,464 @@
+
+Finally, we suggest starting just with builtin resources, as CRDs can be created at
+any time, making it harder to implement with unclear gain. If it appears to not be
+enough (or alternatively we realize that only a subset of builtin resource types is
+needed), this decision can easily be revisited in the future.
+
+#### Reject hanging watches
+
+To mitigate the problem of starving certain PriorityLevels, instead of having the
+watch hang and wait for watchcache initialization, we will simply reject it with
+`Too Many Requests` 429 http code.
+The reason for using this code is that it's already [properly handled] by our client
+libraries (e.g. reflector) and will not risk all of them suddenly falling back
+to list requests. Moreover, given that 429 is a long supported http code, we expect
+that other clients should also properly support it.
+
+[properly handled]: https://github.com/kubernetes/kubernetes/blob/41e706600aea7468f486150d951d3b8948ce89d5/staging/src/k8s.io/client-go/tools/cache/reflector.go#L910-L920
+
+#### Delegate get requests to etcd
+
+Since GET requests have bounded cost, we always estimate their weight the same
+no matter if they are served from etcd or from cache (the latency would be
+different in those cases though, so the actual seat-seconds cost too) and
+finally given we already did some work to process this request, we would
+simply delegate those requests to etcd.
+
+#### Adjust what lists are delegated to etcd
+
+It's tempting to reject all list requests with 429 the same way as watches.
+However, that would also have a negative consequences, e.g. by slowing down
+kube-apiserver initialization (kube-apiserver on startup need to initialize
+its own informers by loop-client, and by rejecting all lists until watchcache
+is initialized, we would effectively block that initialization).
+
+While we didn't decide to reject _all_ list requests with 429, we will start with
+an approach where we only delegate to etcd the requests that (a) are not
+setting any selectors AND (b) are setting a limit.
+The first condition means that every object that we process will also be
+returned, so we're protecting from requests very selective list requests that
+are likely order(s) of magnitude more expensive when served from etcd.
+The second condition effectively bounds the cost of the request.
+
+We may decide to adjust it based on further feedback and/or experiments.
+
+#### Reject the rest of list requests
+
+To mitigate the problem of starving certain PriorityLevels, all other list
+requests will be rejected with 429 http code similarly to watch requests.
+
+
+### Risks and Mitigations
+
+Given we're changing the existing behavior, there is a risk of breaking some
+clients. To mitigate it, we will introduce all the logic behind a feature
+gate, to allow for disabling them if needed.
+
+However, given we're not changing the API itself, we propose to go directly
+to Beta, as Alpha will not give us any feedback.
```

Comment:
Alpha gives the interested parties a chance to test this without impacting default installs (still often used in production).

Reply:
Realistically, providers are not enabling Alpha. There are users who enable Alpha, because they want a given feature, but for changes like this (effectively not being a feature), I doubt anyone would enable it.

That said - are you against the decision of going directly to beta, or was it more of a wording issue with the current text?

Reply:
> That said - are you against the decision of going directly to beta, or was it more of a wording issue with the current text?

I'm against going directly to beta.  I think we should produce an alpha that we can use to see if the feature achieves its goals before we enable such a change on all current default clusters (commonly used in production).